File: earmark-core/channel.go
package earmark

import (
	"crypto/rand"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"time"
)

// ChannelEnvelopeVersion is the current channel message format version.
// Receivers ignore envelopes they do not recognise rather than erroring, so
// bumping this stays backward-compatible.
const ChannelEnvelopeVersion = 1

// ChannelPostMaxAge is how long a channel post is live, measured from its
// posted_at. It matches EarmarkMaxAge: a post is a loan of the sender's stash
// for exactly as long as the sender keeps that stash.
const ChannelPostMaxAge = EarmarkMaxAge

// Envelope types.
const (
	EnvelopeInvite = "invite"
	EnvelopeRoster = "roster"
	EnvelopePost   = "post"
	EnvelopeLeave  = "leave"
)

// Envelope is a channel message. It travels as the content of a NIP-59 rumor,
// so it is only ever seen by one recipient at a time, after two layers of
// decryption. One struct covers all four message types; the type field says
// which fields are meaningful.
type Envelope struct {
	V    int    `json:"v"`
	Chan string `json:"chan"`
	Type string `json:"type"`

	// invite, roster
	Name    string   `json:"name,omitempty"`
	Members []string `json:"members,omitempty"`

	// invite
	Creator   string `json:"creator,omitempty"`
	CreatedAt int64  `json:"created_at,omitempty"`

	// roster
	Seq       int   `json:"seq,omitempty"`
	UpdatedAt int64 `json:"updated_at,omitempty"`

	// post
	PostedAt int64    `json:"posted_at,omitempty"`
	Earmark  *Earmark `json:"earmark,omitempty"`

	// leave
	LeftAt int64 `json:"left_at,omitempty"`
}

// ChannelDescriptor identifies a channel. The ID is 32 random bytes and is
// never derived from the name — it appears only inside encrypted payloads.
type ChannelDescriptor struct {
	ID        string `json:"id"`
	Name      string `json:"name"`
	Creator   string `json:"creator"`
	CreatedAt int64  `json:"created_at"`
}

// Channel is a channel the user has joined, with the roster as they last saw
// it. Seq is the highest roster sequence applied, so a replayed or out-of-order
// roster update can be recognised and dropped.
type Channel struct {
	Descriptor ChannelDescriptor `json:"descriptor"`
	Members    []string          `json:"members"`
	Seq        int               `json:"seq"`
	JoinedAt   int64             `json:"joined_at"`
}

// IsCreator reports whether the given pubkey created this channel, and so may
// change its roster.
func (c *Channel) IsCreator(pubHex string) bool {
	return c.Descriptor.Creator == pubHex
}

// HasMember reports whether pubHex is on the roster.
func (c *Channel) HasMember(pubHex string) bool {
	for _, m := range c.Members {
		if m == pubHex {
			return true
		}
	}
	return false
}

// ChannelInvite is an invite awaiting the user's decision. Trusted records
// whether the sender was in the user's NIP-02 follow list when it arrived —
// it affects how loudly the invite is surfaced and nothing else.
type ChannelInvite struct {
	Descriptor ChannelDescriptor `json:"descriptor"`
	Members    []string          `json:"members"`
	From       string            `json:"from"`
	ReceivedAt int64             `json:"received_at"`
	Trusted    bool              `json:"trusted"`
}

// ChannelPin records chunks this user posted to a channel and must therefore
// keep alive. Purging skips any chunk under a pin younger than
// ChannelPostMaxAge, so cleaning up your own stash cannot break a recipient's
// copy mid-window.
type ChannelPin struct {
	Chan     string   `json:"chan"`
	Chunks   []string `json:"chunks"`
	PostedAt int64    `json:"posted_at"`
}

// ChannelPost is a post received from a channel. Sender is authenticated — it
// is the seal's pubkey, not anything the payload claimed.
type ChannelPost struct {
	Chan     string  `json:"chan"`
	Sender   string  `json:"sender"`
	PostedAt int64   `json:"posted_at"`
	Earmark  Earmark `json:"earmark"`
}

// ID returns a stable identifier for a post, used to dedupe the same message
// arriving from several relays and to address it from the CLI.
func (p *ChannelPost) ID() string {
	return fmt.Sprintf("%s:%s:%d", p.Chan[:8], p.Sender[:8], p.PostedAt)
}

// Expired reports whether the post has aged out of its 30-day window.
func (p *ChannelPost) Expired(now time.Time) bool {
	return p.PostedAt < now.Add(-ChannelPostMaxAge).Unix()
}

// newChannelID returns a fresh 32-byte channel identifier as hex.
func newChannelID() (string, error) {
	b := make([]byte, 32)
	if _, err := rand.Read(b); err != nil {
		return "", fmt.Errorf("could not generate channel id: %w", err)
	}
	return hex.EncodeToString(b), nil
}

func (e *Envelope) encode() ([]byte, error) {
	data, err := json.Marshal(e)
	if err != nil {
		return nil, fmt.Errorf("could not encode channel envelope: %w", err)
	}
	return data, nil
}

func decodeEnvelope(payload []byte) (*Envelope, error) {
	var env Envelope
	if err := json.Unmarshal(payload, &env); err != nil {
		return nil, fmt.Errorf("could not decode channel envelope: %w", err)
	}
	if env.Chan == "" {
		return nil, fmt.Errorf("channel envelope has no channel id")
	}
	return &env, nil
}

File: earmarks-mobile/app/src/main/java/com/derpy/earmarks/data/Channel.kt
package com.derpy.earmarks.data

import org.json.JSONArray
import org.json.JSONObject

/**
 * Channel model, mirroring `earmark-core/channel.go` and `channelstore.go`.
 * Changing any shape here is a protocol change — see the Channels section of
 * docs/PROTOCOL.md.
 */

const val CHANNEL_ENVELOPE_VERSION = 1
const val CHANNEL_STATE_D_TAG = "earmark-channels"

/** A channel post is live for 30 days from its posted_at. */
const val CHANNEL_POST_MAX_AGE_SECONDS = 30L * 24 * 60 * 60

object EnvelopeType {
    const val INVITE = "invite"
    const val ROSTER = "roster"
    const val POST = "post"
    const val LEAVE = "leave"
}

data class ChannelDescriptor(
    val id: String,
    val name: String,
    val creator: String,
    val createdAt: Long
)

data class Channel(
    val descriptor: ChannelDescriptor,
    val members: List<String>,
    val seq: Int,
    val joinedAt: Long
) {
    fun isCreator(pubHex: String) = descriptor.creator == pubHex
    fun hasMember(pubHex: String) = members.contains(pubHex)
}

data class ChannelInvite(
    val descriptor: ChannelDescriptor,
    val members: List<String>,
    val from: String,
    val receivedAt: Long,
    /** Whether the sender was in the user's NIP-02 follow list when it arrived. */
    val trusted: Boolean
)

/** Chunks this user posted and must keep alive for the post's 30-day window. */
data class ChannelPin(
    val chan: String,
    val chunks: List<String>,
    val postedAt: Long
)

/** A post received from a channel. [sender] is authenticated — it is the seal's pubkey. */
data class ChannelPost(
    val chan: String,
    val sender: String,
    val postedAt: Long,
    val earmark: Earmark
) {
    /** Stable id, used to dedupe the same post arriving from several relays. */
    val id: String get() = "${chan.take(8)}:${sender.take(8)}:$postedAt"

    fun expired(nowSeconds: Long) = postedAt < nowSeconds - CHANNEL_POST_MAX_AGE_SECONDS
}

/** Everything the client knows about its channels. Stored self-encrypted. */
data class ChannelState(
    val channels: List<Channel> = emptyList(),
    val invites: List<ChannelInvite> = emptyList(),
    val declined: List<String> = emptyList(),
    val pins: List<ChannelPin> = emptyList()
) {
    fun find(chanId: String): Channel? = channels.firstOrNull { it.descriptor.id == chanId }
    fun findInvite(chanId: String): ChannelInvite? =
        invites.firstOrNull { it.descriptor.id == chanId }

    /** Chunk hashes protected from deletion: those under a pin still in its window. */
    fun pinnedChunks(nowSeconds: Long): Set<String> =
        pins.filter { it.postedAt >= nowSeconds - CHANNEL_POST_MAX_AGE_SECONDS }
            .flatMap { it.chunks }
            .toSet()
}

// --- Serialisation -----------------------------------------------------------
// Field names must match the Go structs exactly; this is the on-the-wire shape.

private fun JSONArray.toStringList(): List<String> =
    (0 until length()).map { getString(it) }

private fun List<String>.toJsonArray(): JSONArray =
    JSONArray().also { arr -> forEach { arr.put(it) } }

fun descriptorFromJson(o: JSONObject) = ChannelDescriptor(
    id = o.getString("id"),
    name = o.optString("name", ""),
    creator = o.optString("creator", ""),
    createdAt = o.optLong("created_at", 0)
)

fun ChannelDescriptor.toJson(): JSONObject = JSONObject().apply {
    put("id", id)
    put("name", name)
    put("creator", creator)
    put("created_at", createdAt)
}

fun channelStateFromJson(json: String): ChannelState {
    val o = JSONObject(json)

    val channels = o.optJSONArray("channels")?.let { arr ->
        (0 until arr.length()).map { i ->
            val c = arr.getJSONObject(i)
            Channel(
                descriptor = descriptorFromJson(c.getJSONObject("descriptor")),
                members = c.optJSONArray("members")?.toStringList() ?: emptyList(),
                seq = c.optInt("seq", 0),
                joinedAt = c.optLong("joined_at", 0)
            )
        }
    } ?: emptyList()

    val invites = o.optJSONArray("invites")?.let { arr ->
        (0 until arr.length()).map { i ->
            val v = arr.getJSONObject(i)
            ChannelInvite(
                descriptor = descriptorFromJson(v.getJSONObject("descriptor")),
                members = v.optJSONArray("members")?.toStringList() ?: emptyList(),
                from = v.optString("from", ""),
                receivedAt = v.optLong("received_at", 0),
                trusted = v.optBoolean("trusted", false)
            )
        }
    } ?: emptyList()

    val pins = o.optJSONArray("pins")?.let { arr ->
        (0 until arr.length()).map { i ->
            val p = arr.getJSONObject(i)
            ChannelPin(
                chan = p.optString("chan", ""),
                chunks = p.optJSONArray("chunks")?.toStringList() ?: emptyList(),
                postedAt = p.optLong("posted_at", 0)
            )
        }
    } ?: emptyList()

    return ChannelState(
        channels = channels,
        invites = invites,
        declined = o.optJSONArray("declined")?.toStringList() ?: emptyList(),
        pins = pins
    )
}

fun ChannelState.toJson(): String = JSONObject().apply {
    put("v", CHANNEL_ENVELOPE_VERSION)
    put("channels", JSONArray().also { arr ->
        channels.forEach { c ->
            arr.put(JSONObject().apply {
                put("descriptor", c.descriptor.toJson())
                put("members", c.members.toJsonArray())
                put("seq", c.seq)
                put("joined_at", c.joinedAt)
            })
        }
    })
    if (invites.isNotEmpty()) put("invites", JSONArray().also { arr ->
        invites.forEach { v ->
            arr.put(JSONObject().apply {
                put("descriptor", v.descriptor.toJson())
                put("members", v.members.toJsonArray())
                put("from", v.from)
                put("received_at", v.receivedAt)
                put("trusted", v.trusted)
            })
        }
    })
    if (declined.isNotEmpty()) put("declined", declined.toJsonArray())
    if (pins.isNotEmpty()) put("pins", JSONArray().also { arr ->
        pins.forEach { p ->
            arr.put(JSONObject().apply {
                put("chan", p.chan)
                put("chunks", p.chunks.toJsonArray())
                put("posted_at", p.postedAt)
            })
        }
    })
}.toString()

/**
 * A decoded channel envelope. One class covers all four message types; [type]
 * says which fields are meaningful.
 */
data class Envelope(
    val v: Int,
    val chan: String,
    val type: String,
    val name: String = "",
    val members: List<String> = emptyList(),
    val creator: String = "",
    val createdAt: Long = 0,
    val seq: Int = 0,
    val updatedAt: Long = 0,
    val postedAt: Long = 0,
    val earmark: Earmark? = null,
    val leftAt: Long = 0
)

/** Returns null for anything unparseable — a malformed envelope is dropped, not raised. */
fun parseEnvelope(json: String): Envelope? = try {
    val o = JSONObject(json)
    val chan = o.optString("chan", "")
    if (chan.isEmpty()) null else Envelope(
        v = o.optInt("v", 0),
        chan = chan,
        type = o.optString("type", ""),
        name = o.optString("name", ""),
        members = o.optJSONArray("members")?.toStringList() ?: emptyList(),
        creator = o.optString("creator", ""),
        createdAt = o.optLong("created_at", 0),
        seq = o.optInt("seq", 0),
        updatedAt = o.optLong("updated_at", 0),
        postedAt = o.optLong("posted_at", 0),
        earmark = o.optJSONObject("earmark")?.let { parseEarmark(it) },
        leftAt = o.optLong("left_at", 0)
    )
} catch (e: Exception) {
    null
}

fun Envelope.toJson(): String = JSONObject().apply {
    put("v", v)
    put("chan", chan)
    put("type", type)
    if (name.isNotEmpty()) put("name", name)
    if (members.isNotEmpty()) put("members", members.toJsonArray())
    if (creator.isNotEmpty()) put("creator", creator)
    if (createdAt != 0L) put("created_at", createdAt)
    if (seq != 0) put("seq", seq)
    if (updatedAt != 0L) put("updated_at", updatedAt)
    if (postedAt != 0L) put("posted_at", postedAt)
    if (earmark != null) put("earmark", earmarkToJsonObject(earmark))
    if (leftAt != 0L) put("left_at", leftAt)
}.toString()
